web/components/roadmap/MiniRoadmap.tsx

```diff
@@ -15,32 +16,35 @@ interface Props {
 
 export const MiniRoadmap = (props: Props): ReactElement => {
   const { roadmap } = useRoadmap();
-  const getActiveStepNumber = useCallback(() => {
-    if (!props.activeTaskId) return undefined;
-    return roadmap?.steps.find((step) => {
-      const currTask = step.tasks.map((it) => it.id);
-      if (props.activeTaskId) return currTask.includes(props.activeTaskId);
-    })?.step_number;
-  }, [props.activeTaskId, roadmap?.steps]);
-  const [activeStepNumber, setActiveStepNumber] = useState<number | undefined>(getActiveStepNumber());
-  const [openSteps, setOpenSteps] = useState<number[]>([]);
-  const { userData } = useUserData();
-  const roadmapExists = !!roadmap;
-
-  useEffect(() => {
-    const activeStep = getActiveStepNumber();
-    setActiveStepNumber(activeStep);
-    setOpenSteps(activeStep ? [activeStep] : []);
-  }, [props.activeTaskId, roadmapExists, getActiveStepNumber]);
+  const { userData, update } = useUserData();
 
-  const toggleStep = (stepNumber: number): void => {
-    analytics.event.task_mini_roadmap_step.click.expand_contract();
-    if (openSteps.includes(stepNumber)) {
-      setOpenSteps(openSteps.filter((number) => number !== stepNumber));
-    } else {
-      setOpenSteps([...openSteps, stepNumber]);
-    }
-  };
+  const toggleStep = useCallback(
+    async (stepNumber: number, isOpen: boolean): Promise<void> => {
+      const updateSteps = (openSteps: number[]) => {
+        return userData
+          ? {
+              ...userData,
+              preferences: {
+                ...userData.preferences,
+                roadmapOpenSteps: openSteps,
+              },
+            }
+          : undefined;
+      };
+      if (!userData) return;
+      const openSteps = userData?.preferences.roadmapOpenSteps;
+      analytics.event.task_mini_roadmap_step.click.expand_contract();
+      if (openSteps.includes(stepNumber)) {
+        if (isOpen) {
+          return;
+        }
```

Comment:
1. is there a refactoring for this nested if?
2. i'm actually confused by this isOpen check... why do we do nothing if the step is open?